### DISCUSSION TITLE
How to prevent JSR from transforming import paths?

### DISCUSSION THREAD

**Discussion by CreepGin** [jsr-io/jsr#576] [Q&A] [answered]
I'm trying to convert some existing repos to be compatible with JSR, but am running into 2 issues.

`jsr publish` is prepending "./" to global module import paths. It's turning
```ts
import { float4 } from "Unity/Mathematics"
```
into 
```ts
import { float4 } from "./Unity/Mathematics"
```

---

Also, it's appending ".ts" to relative import path, automatically turning
```ts
export { palettes } from "./color-palettes"
```
into 
```ts
export { palettes } from "./color-palettes.ts"
```

---

I'm sure there're good reasons for this, but is there any way to prevent it?

**Comment by marvinhagemeister** [comment, marked as answer]
1. That's a bug
2. That's expected. Per specification ESM import specifiers **must** have a file extension at the end.

Creating an issue out of the first one

**Comment by CreepGin** [comment]
@marvinhagemeister Thx! So a follow-up question for 2):

The consumer of my package will get the `An import path can only end with a '.ts' extension when 'allowImportingTsExtensions' is enabled.` error. But enabling `allowImportingTsExtensions` will also require either 'noEmit' or 'emitDeclarationOnly'. 

Looking at the issue here: https://github.com/microsoft/TypeScript/issues/27481

Is [fix-tsc-es-imports](https://www.npmjs.com/package/fix-tsc-es-imports) the only way forward?

**Reply by marvinhagemeister** [reply]
That sounds like a bug. For the npm package we rewrite those `.ts` extensions to something node can understand. Would you mind sharing which package you're testing with?

**Reply by CreepGin** [reply]
Sure, it's https://jsr.io/@onejs/core It's meant to be run inside a custom environment (Unity). The responsible file is `utils/index.ts`. Consumer of the package typically have a tsconfig.json like this:
```json
{
    "compilerOptions": {
        "baseUrl": ".",
        "paths": {
            "*": [
                "*"
            ]
        },
        "target": "esnext",
        "module": "esnext",
        "moduleResolution": "node",
        "sourceMap": true,
        "strictNullChecks": true,
        "skipLibCheck": true,
        "jsx": "react",
        "jsxFactory": "h",
        "jsxFragmentFactory": "Fragment",
        "lib": [
            "esnext"
        ],
        "typeRoots": [
            "./node_modules/@types"
        ],
        "outDir": "@outputs/tsc"
    }
}
```

**Comment by HuakunShen** [comment]
@CreepGin Did you find a solution? I also had this problem. The issue was not closed https://github.com/jsr-io/jsr/issues/577

**Reply by CreepGin** [reply]
Sorry nope. I had to forgo JSR due to other issues. Haven't had a chance to give it another go yet. Issues I was waiting on:

https://github.com/jsr-io/jsr/issues/24
https://github.com/jsr-io/jsr/issues/339

**Reply by HuakunShen** [reply]
Thanks, I also had these issues. Currently I only use jsr for pure TypeScript library, had to use npm for component lib and other more complicated things.
